File: backend/app/schemas/agent.py
from typing import List, Optional, Dict, Any, Union
from pydantic import BaseModel, Field

class AgentContext(BaseModel):
    project_id: Optional[int] = None
    chapter_id: Optional[str] = None
    scene_id: Optional[str] = None
    selected_text: Optional[str] = None
    language: Optional[str] = "zh"

class AgentRequest(BaseModel):
    message: str
    context: AgentContext
    history: List[Dict[str, str]] = [] # [{"role": "user", "content": "..."}, ...]

class ToolCall(BaseModel):
    tool_name: str
    arguments: Dict[str, Any]
    reason: str

class AgentResponse(BaseModel):
    thought: str = Field(..., description="The internal reasoning process of the agent.")
    response: str = Field(..., description="The natural language response to the user.")
    action: Optional[ToolCall] = Field(None, description="An optional tool call to execute.")

class ToolResult(BaseModel):
    tool_name: str
    result: Any
    status: str = "success" # success, error

File: backend/src/schemas/agent.ts
import { z } from 'zod';

export const AgentContextSchema = z.object({
  project_id: z.number().int().optional().nullable(),
  chapter_id: z.string().optional().nullable(),
  scene_id: z.string().optional().nullable(),
  selected_text: z.string().optional().nullable(),
  language: z.string().default('zh').optional().nullable(),
  /** Current app route hint, e.g. story | director | characters */
  route: z.string().optional().nullable(),
});

export const AgentRequestSchema = z.object({
  message: z.string(),
  context: AgentContextSchema,
  history: z.array(z.record(z.string(), z.any())).default([]),
});

export const ToolCallSchema = z.object({
  tool_name: z.string(),
  arguments: z.record(z.string(), z.any()),
  reason: z.string().optional(),
});

export const AgentResponseSchema = z.object({
  thought: z.string(),
  response: z.string(),
  /** Agent OS multi-action plan */
  actions: z.array(z.record(z.string(), z.any())).optional().default([]),
  needs_confirmation: z.boolean().optional().default(false),
  /** @deprecated legacy single tool call for old UI */
  action: ToolCallSchema.optional().nullable(),
});

export const ToolResultSchema = z.object({
  tool_name: z.string(),
  result: z.any(),
  status: z.string().default('success'),
});

export type AgentContext = z.infer<typeof AgentContextSchema>;
export type AgentRequest = z.infer<typeof AgentRequestSchema>;
export type ToolCall = z.infer<typeof ToolCallSchema>;
export type AgentResponse = z.infer<typeof AgentResponseSchema>;
export type ToolResult = z.infer<typeof ToolResultSchema>;
